source/h2_frames.c

```diff
@@ -332,714 +256,829 @@ void aws_h2_frame_encoder_clean_up(struct aws_h2_frame_encoder *encoder) {
+        frame_type = frame->base.type;
+
+        if (frame->pad_length > 0) {
+            flags |= AWS_H2_FRAME_F_PADDED;
+            pad_length = frame->pad_length;
+            payload_overhead += 1 + pad_length;
         }
-        flags |= AWS_H2_FRAME_F_ACK;
-    }
 
-    if (s_frame_prefix_encode(&frame->base, frame->settings_count * 6, flags, output)) {
-        goto write_error;
-    }
+        if (frame->has_priority) {
+            priority_settings = &frame->priority;
+            flags |= AWS_H2_FRAME_F_PRIORITY;
+            payload_overhead += s_frame_priority_settings_size;
+        }
 
-    /* Write the payload */
-    for (size_t i = 0; i < frame->settings_count; ++i) {
-        if (!aws_byte_buf_write_be16(output, frame->settings_array[i].id) ||
-            !aws_byte_buf_write_be32(output, frame->settings_array[i].value)) {
+        if (frame->end_stream) {
+            flags |= AWS_H2_FRAME_F_END_STREAM;
+        }
 
-            aws_raise_error(AWS_ERROR_SHORT_BUFFER);
-            goto write_error;
+        if (frame_type == AWS_H2_FRAME_T_PUSH_PROMISE) {
+            promised_stream_id = &frame->promised_stream_id;
+            payload_overhead += 4;
         }
-    }
 
-    return AWS_OP_SUCCESS;
+    } else /* CONTINUATION */ {
+        frame_type = AWS_H2_FRAME_T_CONTINUATION;
+    }
 
-write_error:
-    output->len = output_init_len;
-    return AWS_OP_ERR;
-}
+    /*
+     * Figure out what size header-block fragment should go in this frame.
+     */
 
-/***********************************************************************************************************************
- * PUSH_PROMISE
- **********************************************************************************************************************/
-int aws_h2_frame_push_promise_init(struct aws_h2_frame_push_promise *frame, struct aws_allocator *allocator) {
-    (void)allocator;
+    size_t max_payload;
+    if (s_get_max_contiguous_payload_length(encoder, output, &max_payload)) {
+        goto handle_waiting_for_more_space;
+    }
 
-    AWS_ZERO_STRUCT(*frame);
-    frame->base.type = AWS_H2_FRAME_T_PUSH_PROMISE;
+    size_t max_fragment;
+    if (aws_sub_size_checked(max_payload, payload_overhead, &max_fragment)) {
+        goto handle_waiting_for_more_space;
+    }
 
-    return aws_h2_frame_header_block_init(&frame->header_block, allocator);
-}
-void aws_h2_frame_push_promise_clean_up(struct aws_h2_frame_push_promise *frame) {
-    AWS_PRECONDITION(frame);
+    const size_t fragment_len = aws_min_size(max_fragment, frame->header_block_cursor.len);
+    if (fragment_len == frame->header_block_cursor.len) {
+        /* This will finish the header-block */
+        flags |= AWS_H2_FRAME_F_END_HEADERS;
+    } else {
+        /* If we're not finishing the header-block, is it even worth trying to send this frame now? */
+        const size_t even_worth_sending_threshold = s_frame_prefix_length + payload_overhead;
```

Comment:
we could probably make this a bit smarter in the future. Currently this would just be, do we at least have enough space to send a valid but empty headers frame?

Reply:
Oh, I can easily do that.
I was trying to avoid sending a 9-byte HEADERS frame that had no actual header-data in it, knowing that the next time we try to encode there will be a fresh new aws_io_message with tons of space.

Do you think it's valuable to send the valid-but-empty frame earlier?